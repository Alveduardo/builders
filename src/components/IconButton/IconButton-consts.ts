export const ICON_BUTTON = {
  TEST_ID: {
    CONTAINER: 'ICON_BUTTON_CONTAINER_TEST_ID',
    ACTIVITY_INDICATOR: 'ICON_BUTTON_ACTIVITY_INDICATOR_TEST_ID',
    ICON: 'ICON_BUTTON_ICON_TEST_ID',
  },
};
