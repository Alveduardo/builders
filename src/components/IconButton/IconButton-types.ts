import { ActivityIndicatorProps, TouchableHighlightProps, ViewStyle } from "react-native";
import { Icon, IconProps } from "react-native-vector-icons/Icon";

export type Size = 'small' | 'regular' | 'large'

export type SizeStyles = {
    [key in Size]: {
        containerStyle: ViewStyle
        sizeIcon: number
    }
}

interface DefaultStyles {
    containerStyle: ViewStyle
}

export interface IconButtonStylesType {
    defaultStyles: DefaultStyles,
    sizeStyles: SizeStyles
}

interface StateConfig {
    iconProps: IconProps
    sizeIndicator: ActivityIndicatorProps['size']
}

type StylesConfig = DefaultStyles

export interface IconButtonConfig {
    state: StateConfig
    styles: StylesConfig
}

export interface IconButtonPropsType extends Omit<IconProps, 'size' | 'style'>, TouchableHighlightProps {
    type: typeof Icon,
    size: Size
    loading?: boolean
}