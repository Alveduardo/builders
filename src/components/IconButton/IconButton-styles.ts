import { IconButtonStylesType } from "./IconButton-types";

export const IconButtonStyles: IconButtonStylesType = {
    defaultStyles: {
        containerStyle: {
            alignItems: 'center',
            justifyContent: 'center',
        }
    },
    sizeStyles: {
        small: {
            containerStyle: {
                width: 32 + 24,
                height: 32 + 24,
            },
            sizeIcon: 32
        },
        regular: {
            containerStyle: {
                width: 40 + 30,
                height: 40 + 30,
            },
            sizeIcon: 48
        },
        large: {
            containerStyle: {
                width: 64 + 36,
                height: 64 + 36,
            },
            sizeIcon: 64
        }
    }
} 