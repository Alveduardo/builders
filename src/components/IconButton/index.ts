import IconButton from './IconButton';
import { IconButtonProps } from './IconButton-types';

export default IconButton;
export { IconButtonProps };
