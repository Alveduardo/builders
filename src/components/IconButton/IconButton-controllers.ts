import { iconButtonStyles } from './IconButton-styles';
import { IconButtonConfig, IconButtonProps } from './IconButton-types';

export const useIconButtonConfig = ({ name, size, color }: Partial<IconButtonProps>): IconButtonConfig => {
  const {
    defaultStyles: { containerStyle },
    sizeStyles: {
      [size!]: { iconSize },
    },
  } = iconButtonStyles;

  const sizeIndicator = size === 'large' ? 'large' : 'small';

  return {
    state: {
      iconProps: {
        name: name!,
        size: iconSize,
        color,
      },
      sizeIndicator,
    },
    styles: {
      containerStyle,
    },
  };
};
