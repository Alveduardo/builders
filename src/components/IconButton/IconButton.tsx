import React, { memo } from 'react';
import { ActivityIndicator, ColorValue, TouchableOpacity } from 'react-native';
import { areEqualComponentProps } from '../../utils';
import { ICON_BUTTON } from './IconButton-consts';
import { useIconButtonConfig } from './IconButton-controllers';
import { IconButtonProps } from './IconButton-types';

const IconButton = memo(
  ({
    type,
    name,
    size = 'small',
    color = '#FFF',
    loading = false,
    style = undefined,
    ...rest
  }: IconButtonProps): JSX.Element => {
    const {
      state: { iconProps, sizeIndicator },
      styles: { containerStyle },
    } = useIconButtonConfig({
      name,
      size,
      color,
    });

    const Icon = type;

    return (
      <TouchableOpacity testID={ICON_BUTTON.TEST_ID.ICON} {...rest} style={[containerStyle, style]}>
        {loading ? (
          <ActivityIndicator
            testID={ICON_BUTTON.TEST_ID.ACTIVITY_INDICATOR}
            size={sizeIndicator}
            color={color as ColorValue}
          />
        ) : (
          <Icon testID={ICON_BUTTON.TEST_ID.ICON} {...iconProps} />
        )}
      </TouchableOpacity>
    );
  },
  areEqualComponentProps,
);

export default IconButton;
