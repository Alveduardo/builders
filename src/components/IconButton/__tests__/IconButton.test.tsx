import React from 'react';
import { render } from '@testing-library/react-native';
import IconButton from '../IconButton';
import AntDesign from 'react-native-vector-icons/AntDesign';
import { StyleSheet } from 'react-native';
import { ICON_BUTTON } from '../IconButton-consts';
import { Size as IconSize } from '../IconButton-types';

const iconButtonPropsDefault = {
  type: AntDesign,
  name: 'home',
};

const describeEachLoading = [
  {
    loading: false,
  },
  {
    loading: true,
  },
];

const describeEachSize = [
  {
    size: 'xxxSmall',
    iconSize: 16,
    sizeIndicator: 'small',
  },
  {
    size: 'xxSmall',
    iconSize: 20,
    sizeIndicator: 'small',
  },
  {
    size: 'xSmall',
    iconSize: 28,
    sizeIndicator: 'small',
  },
  {
    size: 'small',
    iconSize: 32,
    sizeIndicator: 'small',
  },
  {
    size: 'regular',
    iconSize: 48,
    sizeIndicator: 'small',
  },
  {
    size: 'large',
    iconSize: 64,
    sizeIndicator: 'large',
  },
];

describe('IconButton ->', () => {
  beforeEach(() => {
    jest.spyOn(console, 'error').mockImplementation(jest.fn());
  });
  afterEach(() => {
    jest.clearAllMocks();
  });
  afterAll(() => {
    jest.resetAllMocks();
  });

  describe('Snapshot ->', () => {
    it('Deve corresponder ao componente com as props default', () => {
      const tree = render(<IconButton {...iconButtonPropsDefault} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('Deve corresponder ao componente com as props mais utilizadas', () => {
      const tree = render(<IconButton {...iconButtonPropsDefault} color={'#FF0000'} size={'regular'} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  describe('props ->', () => {
    describe('loading ->', () => {
      describe.each(describeEachLoading)('Para cada valor de loading ->', ({ loading }) => {
        it(`Deve renderizar o componente de acordo com a prop ${loading}`, () => {
          const { queryByTestId } = render(<IconButton {...iconButtonPropsDefault} {...{ loading }} />);

          const componentActivityIndicator = queryByTestId(ICON_BUTTON.TEST_ID.ACTIVITY_INDICATOR);
          const componentIcon = queryByTestId(ICON_BUTTON.TEST_ID.ICON);

          if (loading) {
            expect(componentActivityIndicator).toBeDefined();
            expect(componentIcon).toBeNull();
          } else {
            expect(componentActivityIndicator).toBeNull();
            expect(componentIcon).toBeDefined();
          }
        });
      });
    });
    describe('color ->', () => {
      it('Deve renderizar o componente alterando a prop color', () => {
        const color = '#FF0000';

        const { getByTestId } = render(<IconButton {...iconButtonPropsDefault} {...{ color }} />);

        const containerStyle = getByTestId(ICON_BUTTON.TEST_ID.ICON).props.style;

        expect(containerStyle[0].color).toBe(color);
      });
    });
    describe('size ->', () => {
      describe.each(describeEachSize)('Para cada valor de size ->', ({ size, iconSize, sizeIndicator }) => {
        it(`Deve ter o valor de ${iconSize} de acordo com a prop ${size}`, () => {
          const { getByTestId } = render(<IconButton {...iconButtonPropsDefault} size={size as IconSize} />);

          const containerStyle = getByTestId(ICON_BUTTON.TEST_ID.ICON).props.style;

          expect(containerStyle[0].fontSize).toBe(iconSize);
        });
        it(`Deve ter o valor de ${sizeIndicator} de acordo com a prop ${size}`, () => {
          const { getByTestId } = render(<IconButton {...iconButtonPropsDefault} size={size as IconSize} loading />);

          const containerStyle = StyleSheet.flatten(getByTestId(ICON_BUTTON.TEST_ID.ACTIVITY_INDICATOR).props);

          expect(containerStyle.size).toBe(sizeIndicator);
        });
      });
    });
  });
});
