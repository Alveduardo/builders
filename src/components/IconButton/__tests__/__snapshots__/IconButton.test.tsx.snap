// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconButton -> Snapshot -> Deve corresponder ao componente com as props default 1`] = `
<View
  disabled={true}
  rest={Object {}}
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
    }
  }
  testID="ICON_BUTTON_CONTAINER_TEST_ID"
>
  <Text
    allowFontScaling={false}
    selectable={false}
    style={
      Array [
        Object {
          "color": "#FFF",
          "fontSize": 32,
        },
        undefined,
        Object {
          "fontFamily": "anticon",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
    testID="ICON_BUTTON_ICON_TEST_ID"
  >
    
  </Text>
</View>
`;

exports[`IconButton -> Snapshot -> Deve corresponder ao componente com as props mais utilizadas 1`] = `
<View
  disabled={true}
  rest={Object {}}
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
    }
  }
  testID="ICON_BUTTON_CONTAINER_TEST_ID"
>
  <Text
    allowFontScaling={false}
    selectable={false}
    style={
      Array [
        Object {
          "color": "#FF0000",
          "fontSize": 48,
        },
        undefined,
        Object {
          "fontFamily": "anticon",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
    testID="ICON_BUTTON_ICON_TEST_ID"
  >
    
  </Text>
</View>
`;
